test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	variables := map[string]string{}
+	ipamService := newIPAMService(e2eIPAMKubeconfig, skipCleanup)
+
+	// Claim an IP per variable.
+	for _, variable := range append(additionalIPVariableNames, controlPlaneEndpointVariable) {
+		ip := ipamService.claimIPAddress(ctx)
+		Byf("Setting clusterctl variable %s to %s", variable, ip)
+		variables[variable] = ip
+	}
+
+	// Create a new clusterctl config file based on the passed file and add the new variables for the IPs.
+	modifiedClusterctlConfigPath := fmt.Sprintf("%s-%s.yaml", strings.TrimSuffix(clusterctlConfigPath, ".yaml"), rand.String(16))
+	Byf("Writing a new clusterctl config to %s", modifiedClusterctlConfigPath)
+	copyAndAmendClusterctlConfig(ctx, copyAndAmendClusterctlConfigInput{
+		ClusterctlConfigPath: clusterctlConfigPath,
+		OutputPath:           modifiedClusterctlConfigPath,
+		Variables:            variables,
+	})
+
+	return modifiedClusterctlConfigPath, ipamService.cleanup
+}
+
+type ipamService struct {
+	client          client.Client
+	ipAddressClaims []ipamv1.IPAddressClaim
+	skipCleanup     bool
+}
+
+func newIPAMService(e2eIPAMKubeconfig string, skipCleanup bool) *ipamService {
+	kubeConfig, err := os.ReadFile(filepath.Clean(e2eIPAMKubeconfig))
+	Expect(err).ToNot(HaveOccurred())
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfig)
+	Expect(err).ToNot(HaveOccurred())
+
+	ipamClient, err := client.New(restConfig, client.Options{Scheme: ipamScheme})
+	Expect(err).ToNot(HaveOccurred())
+
+	return &ipamService{
+		ipAddressClaims: []ipamv1.IPAddressClaim{},
+		client:          ipamClient,
+		skipCleanup:     skipCleanup,
+	}
+}
+
+func (s *ipamService) claimIPAddress(ctx context.Context) string {
+	claim := &ipamv1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "ipclaim-" + rand.String(32),
+			Namespace: metav1.NamespaceDefault,
+			Annotations: map[string]string{
+				"prow.k8s.io/build-id": os.Getenv("BUILD_ID"),
+				"prow.k8s.io/job":      os.Getenv("JOB_NAME"),
+			},
+		},
+		Spec: ipamv1.IPAddressClaimSpec{
+			PoolRef: corev1.TypedLocalObjectReference{
+				APIGroup: ptr.To("ipam.cluster.x-k8s.io"),
+				Kind:     "InClusterIPPool",
+				Name:     "capv-e2e-ippool",
+			},
+		},
+	}
+
+	// Create an IPAddressClaim
+	Byf("Creating IPAddressClaim %s/%s", claim.GetNamespace(), claim.GetName())
+	Expect(s.client.Create(ctx, claim)).To(Succeed())
+
+	// Wait for the IPAddressClaim to refer an IPAddress.
+	Eventually(func() error {
+		if err := s.client.Get(ctx, client.ObjectKeyFromObject(claim), claim); err != nil {
+			return err
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			return fmt.Errorf("IPAddressClaim does not yet have an IP")
+		}
+
+		return nil
+	}, time.Second*30, time.Second).Should(Succeed())
+
+	ip := &ipamv1.IPAddress{}
+
+	Expect(s.client.Get(ctx, client.ObjectKey{Namespace: claim.GetNamespace(), Name: claim.Status.AddressRef.Name}, ip)).To(Succeed())
+	Expect(ip.Spec.Address).ToNot(Equal(""))
+
```

Comment:
With these expects we will leak any resources if the Expect fails, correct?

Would it help to return an error instead that we check one level higher?

Or maybe we get the same result if we move l.151 up?